### PR TITLE
Add ruff for shared mutable defaults (B)

### PR DESCRIPTION
**Pull request by enyst**
This PR proposes to add to pre-commit checks:
- B006 [mutable-argument-default](https://docs.astral.sh/ruff/rules/mutable-argument-default/)
- B008 [function-call-in-default-argument](https://docs.astral.sh/ruff/rules/function-call-in-default-argument/)
- I ran on all files locally and included fixes for reported issues (except fast api `Depends` argument, which is not obviously wrong to me)

Additionally, it adds most of "B", minus a few that may be a bit annoying and/or less relevant. If we don't want more, perhaps we can consider only B006 and B008.

It's a very counter-intuitive python gotcha. ~~Addresses https://github.com/OpenDevin/OpenDevin/pull/1910#discussion_r1606940945~~
Addresses the root cause of https://github.com/OpenDevin/OpenDevin/pull/1713

Example result CI on opendevin directory:
https://github.com/enyst/OpenDevin/actions/runs/9171750274/job/25216761272?pr=3

Feedback appreciated!

Reference: https://docs.astral.sh/ruff/rules/#flake8-bugbear-b

### CODE REVIEW

**Comment by li-boxuan** [comment]
I am a little bit confused about how this PR addresses https://github.com/OpenDevin/OpenDevin/pull/1910#discussion_r1606940945. Do you mind elaborating?

File: agenthub/SWE_agent/prompts.py [hunk 1]
```diff
@@ -191,7 +192,7 @@ def MEMORY_FORMAT(act, obs):
 {unpack_dict(act, ["content"])}
 
 Output from Action:
-{unpack_dict(obs)}
+{unpack_dict(obs), ["content"]}
```
Comment (neubig) [inline agenthub/SWE_agent/prompts.py:195]: Two comments:
1. This changes the program behavior, is that intentional?
2. If it is intentional, maybe there's a mis-placed parenthesis?

```suggestion
{unpack_dict(obs, ["content"])}
```

**Comment by enyst** [comment]
@li-boxuan 

> I am a little bit confused about how this PR addresses [#1910 (comment)](https://github.com/OpenDevin/OpenDevin/pull/1910#discussion_r1606940945). Do you mind elaborating?

If we introduce rules like B006 and B008 pre-commit, then we address issues like this:

https://github.com/OpenDevin/OpenDevin/pull/1713

before they happen. I really think we should do this, even if take in only these two rules.

I can see why the relationship with that comment is not clear, wow, my bad. Fixing the reference now.

What I mean is: the issues caused by mutable defaults can be annoying or serious enough to lead to over correction in other parts of the code. I think 1713 tilted the balance towards needing to raise an exception [here](https://github.com/OpenDevin/OpenDevin/pull/1910/files/9c719cf25a3b4fb2004c0091023d5b0126e88526#diff-081383e511233feb06a2eaef61d764a8543aa7b8229e1352ac05eacc74400233L67).

To be clear, I'm just guessing, I don't think Robert said that explicitly. It's just my opinion about why the code is shaping up in a particular way, and it's why I didn't comment on it.

Allow me to take another example: the custom `llm_config` feature you suggested has proven amazingly useful in ways I didn't realize at the time, and I'm so glad you brought it out! During its implementation, I frankly thought it was going to be easy, and it ended up taking far more time than I had possibly guessed. Why? It's a bit embarrassing to admit, I wasted hours because of this python behavior.

The reason there was that unlike other config-related things, this one has to happen after everything was set up, and after a bunch of other imports happen. Those imports initialize damn parameters... with values that never change when you execute the call.

`llmconfig` was updated before _execution_ of function calls, with a custom group of settings, but function calls were never receiving the update. And I didn't see it. I suspected everything else on the execution path, reasonable or otherwise, except this Python behavior...

This behavior is as old as Python. Now this Ruff automation won't solve every kind of parameter passing in those issues I've had, but it will solve some. And it will attract our attention, reminding us it's happening. I believe we should automate it, instead of working around or wasting more time anywhere else.

File: agenthub/SWE_agent/prompts.py (hunk 1 above)
Comment (enyst) [reply agenthub/SWE_agent/prompts.py]: Re: 1. You're right, I just verified, this changed the previous behavior and I don't think it should. Fixed now.

**Review by neubig — COMMENTED** [review]
@enyst I think this looks good, but it looks like we need to fix integration tests.

**Comment by li-boxuan** [comment]
Oh @enyst thanks for the explanation! Damn I never knew this python behavior lol... I've been mostly a java and C++ developer so this is totally something I never imagined.

**Review by li-boxuan — APPROVED** [review]

**Comment by li-boxuan** [comment]
Pulling main once more before we merge this PR...

**Pull request merged**